keras_nlp/metrics/edit_distance.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Edit Distance metric implementation based on `keras.metrics.Metric`."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+class EditDistance(keras.metrics.Metric):
+    """Edit Distance.
+
+    This class implements the edit distance metric, sometimes called
+    Levenshtein Distance. Essentially, edit distance is the least number of
+    operations required to convert one string to another, where a operation
+    can be one of substitution, deletion or insertion. In the normalized version,
+    the above score is divided by the number of tokens in the reference text.
+
+    This class can be used to compute character error rate (CER)/word error rate
+    (WER). You simply have to pass the appropriate tokenized text, and set
+    `normalize` to True.
+
+    Note on input shapes:
+    `y_true` and `y_pred` can either be tensors of rank 1 or ragged tensors of
+    rank 2. These tensors contain tokenized text.
+
+    Args:
+        normalize: bool. If True, the computed number of operations
+            (substitutions + deletions + insertions) across all samples is
+            divided by the aggregate number of tokens in all reference texts. If
+            False, number of operations are calculated for every sample, and
+            averaged over all the samples.
+        dtype: string or tf.dtypes.Dtype. Precision of metric computation. If
+            not specified, it defaults to tf.float32.
+        name: string. Name of the metric instance.
+        **kwargs: Other keyword arguments.
+
+    References:
+        - [Morris et al.](https://www.researchgate.net/publication/221478089_From_WER_and_RIL_to_MER_and_WIL_improved_evaluation_measures_for_connected_speech_recognition)
```

Comment:
It looks like https://www.researchgate.net/publication/221478089 works as a slightly more abbreviated link